protocols/kad/src/protocol.rs

```diff
@@ -179,6 +177,47 @@ impl UpgradeInfo for KademliaProtocolConfig {
     }
 }
 
+/// Codec for Kademlia inbound and outbound message framing.
+pub struct KadCodec<A, B> {
+    codec: quick_protobuf_codec::Codec<proto::Message>,
+    __phantom: PhantomData<(A, B)>,
+}
+impl<A, B> KadCodec<A, B> {
+    fn new(max_packet_size: usize) -> Self {
+        KadCodec {
+            codec: quick_protobuf_codec::Codec::new(max_packet_size),
+            __phantom: PhantomData,
+        }
+    }
+}
+
+impl<A: Into<proto::Message>, B> Encoder for KadCodec<A, B> {
+    type Error = io::Error;
+    type Item = A;
+
+    fn encode(&mut self, item: Self::Item, dst: &mut BytesMut) -> Result<(), Self::Error> {
+        self.codec
+            .encode(item.into(), dst)
+            .map_err(|err| err.into())
+    }
+}
+impl<A, B: TryFrom<proto::Message, Error = io::Error>> Decoder for KadCodec<A, B> {
+    type Error = io::Error;
+    type Item = B;
+
+    fn decode(&mut self, src: &mut BytesMut) -> Result<Option<Self::Item>, Self::Error> {
+        self.codec
+            .decode(src)?
+            .map(|msg| B::try_from(msg))
+            .transpose()
+    }
+}
+
+/// Sink of responses and stream of requests.
+pub(crate) type KadInStreamSink<S> = Framed<S, KadCodec<KadResponseMsg, KadRequestMsg>>;
+/// Sink of requests and stream of responses.
+pub(crate) type KadOutStreamSink<S> = Framed<S, KadCodec<KadRequestMsg, KadResponseMsg>>;
+
```

Comment:
Are these still useful now? Or should we just delete them?

Reply:
Not necessarily, but it does prevent us having to write the more lengthy `Framed<Stream, Codec<KadResponseMsg, KadRequestMsg>>` multiple times in `handler.rs`

Reply:
Okay, happy to keep them then!